ethcore/src/client/client.rs

```diff
@@ -1551,23 +1556,35 @@ impl BlockChainClient for Client {
 	}
 
 	fn replay_block_transactions(&self, block: BlockId, analytics: CallAnalytics) -> Result<Box<Iterator<Item = Executed>>, CallError> {
-		let mut env_info = self.env_info(block).ok_or(CallError::StatePruned)?;
-		let body = self.block_body(block).ok_or(CallError::StatePruned)?;
-		let mut state = self.state_at_beginning(block).ok_or(CallError::StatePruned)?;
-		let txs = body.transactions();
-		let engine = self.engine.clone();
+		let block_hash = self.block_hash(block).ok_or(CallError::StatePruned)?;
+		let data_cached = {
```

Comment:
Instead of acquiring the lock twice, better do this:
```rust
if let Some(cached) = self.replay_cache.lock().get_mut(&block_hash) {
  return Ok(Box::new(cached.clone().into_iter()));
}
// whatever should be in else block
//...
```